### ISSUE TITLE
Class-qualified annotation expressions

### ISSUE THREAD

**Issue by samwgoldman** [python/typing#1884]
The wonderfully useful [specification of type annotation expressions](https://typing.readthedocs.io/en/latest/spec/annotations.html#type-and-annotation-expressions) describes what kinds of expressions are valid type annotations.

I believe the spec does not include a case which type checkers support, although as we will see this comes down to semantics :)

```
class A:
    class B:
        pass

x: A.B = A.B()
```

The `A.B` annotation seems to fall under the `name` rule:

```
| name
      (where name must refer to a valid in-scope class,
       type alias, or TypeVar)
```

The spec helpfully continues:

```
Any leaf denoted as name may also be a qualified name (i.e., module '.' name or package '.' module '.' name, with any level of nesting).
```

The concept of a "qualified name" occurs only in this section, and the parenthetical explanation serves as it's most precise definition, as far as I can tell.

Should we extend the definition of "qualified name" to also include class members? Or maybe I should read the class as a "module" or a "package" and the existing definition suffices?

**Comment by erictraut** [comment]
> Should we extend the definition of "qualified name" to also include class members?

Yes, I think that would be a good clarification. This section focuses primarily on the *syntactic* rules for type expressions and intentionally omits some of the detailed *semantic* rules because they are covered in more detail elsewhere in the spec. That said, I agree that there's room for additional clarity in the notes section here.

**Comment by JelleZijlstra** [comment]
Yes, this is a good point; I overlooked this possibility when I wrote that text.

**Comment by carljm** [comment]
In terms of type checker support, [mypy does not seem to support](https://mypy-play.net/?mypy=latest&python=3.12&gist=8b24d52d4911a15042b4f11b50615aed) class-scoped type aliases:

```py
class A:
    X = int

def f(x: A.X):
    reveal_type(x)
```

**Comment by hauntsaninja** [comment]
(I think mypy does support them with PEP 695 or 613 aliases)
